### PR TITLE
Add IR support for Wasm globals

### PR DESCRIPTION
**Pull request by gkaracha**
This PR shall add support for global variables in Asterius' PR. Currently work in progress.

### CODE REVIEW

File: asterius/src/Asterius/Types.hs [hunk 1]
```diff
@@ -505,6 +557,9 @@ data Module
         functionTable :: FunctionTable,
         tableImport :: TableImport,
         tableSlots :: Int,
+        globalImports :: [GlobalImport],
+        globalExports :: [GlobalExport],
+        globalMap :: [Global],
```
Comment (TerrorJack) [inline asterius/src/Asterius/Types.hs:562]: This should be a `SymbolMap Global`.

File: asterius/src/Asterius/Types.hs [hunk 2]
```diff
@@ -450,6 +478,17 @@ data Expression
...
   deriving (Show, Data)
 
+-- | A 'Global' captures a single global variable. Each 'Global' stores a
+-- single value of the given global type.  Moreover, each 'Global' is
+-- initialized with an initial value, given by a constant initializer
+-- expression.
+data Global
+  = Global
+      { globalType :: GlobalType,
+        init :: Expression
```
Comment (TerrorJack) [inline asterius/src/Asterius/Types.hs:488]: Might as well call this field `globalInit`.

File: asterius/src/Asterius/Types.hs [hunk 3]
```diff
@@ -345,6 +350,21 @@ data BinaryOp
   | GeFloat64
   deriving (Show, Data)
 
+-- | 'Mutability' of variables.
+data Mutability
+  = Const  -- ^ Immutable.
```
Comment (TerrorJack) [inline asterius/src/Asterius/Types.hs:355]: `Const` exists in `base`, and `Var` exists in GHC API; although the name collision is not likely, I'd still suggest just naming the datacons as `Immutable`/`Mutable` here.

Comment (gkaracha) [reply asterius/src/Asterius/Types.hs:355]: Actually I came across a collision already (in `Asterius.Foreign.DsForeign`)! I used Var/Const to match the Wasm spec but I like Mutable/Immutable better :slightly_smiling_face:

File: asterius/src/Asterius/Resolve.hs [hunk 4]
```diff
@@ -93,6 +93,9 @@ resolveAsteriusModule debug bundled_ffi_state m_globals_resolved func_start_addr
             externalBaseName = "table"
           },
         tableSlots = table_slots,
+        globalImports = mempty, -- TODO
```
[the remark this answers is not in the document]
Comment (gkaracha) [reply asterius/src/Asterius/Resolve.hs:96]: I see. Currently it'd still be empty though, right? Something like this maybe:
```diff
diff --git a/asterius/src/Asterius/Builtins.hs b/asterius/src/Asterius/Builtins.hs
index 3eb3a0ca..c1d5228a 100644
--- a/asterius/src/Asterius/Builtins.hs
+++ b/asterius/src/Asterius/Builtins.hs
@@ -8,6 +8,8 @@ module Asterius.Builtins
     rtsAsteriusModule,
     rtsFunctionImports,
     rtsFunctionExports,
+    rtsGlobalImports,
+    rtsGlobalExports,
     emitErrorMessage,
     wasmPageSize,
     generateWrapperFunction,
@@ -681,6 +683,12 @@ rtsFunctionExports debug =
       }
   ]
 
+rtsGlobalImports :: [GlobalImport]
+rtsGlobalImports = mempty
+
+rtsGlobalExports :: [GlobalExport]
+rtsGlobalExports = mempty
+
 emitErrorMessage :: [ValueType] -> BS.ByteString -> Expression
 emitErrorMessage vts ev = Barf {barfMessage = ev, barfReturnTypes = vts}
 
diff --git a/asterius/src/Asterius/Resolve.hs b/asterius/src/Asterius/Resolve.hs
index dbb0159f..1bf7bea1 100644
--- a/asterius/src/Asterius/Resolve.hs
+++ b/asterius/src/Asterius/Resolve.hs
@@ -93,8 +93,8 @@ resolveAsteriusModule debug bundled_ffi_state m_globals_resolved func_start_addr
             externalBaseName = "table"
           },
         tableSlots = table_slots,
-        globalImports = mempty, -- TODO
-        globalExports = mempty, -- TODO
+        globalImports = rtsGlobalImports,
+        globalExports = rtsGlobalExports,
         globalMap = globalsMap m_globals_resolved, -- Copy as-is.
         memorySegments = segs,
         memoryImport = MemoryImport
```

[the remark this answers is not in the document]
Comment (TerrorJack) [reply asterius/src/Asterius/Resolve.hs:96]: Yes, for this PR it'll be empty. We'll add entries to `rtsGlobalImports`/`rtsGlobalExports` when we need them in later PRs

**Comment by gkaracha** [comment]
@TerrorJack this should be ready for review now. I tried to test its correctness with PR #731 but I am not entirely sure if it worked; CI is failing very frequently.

**Pull request merged**